modules/photo/src/ccm/ccm.cpp

```diff
@@ -0,0 +1,504 @@
+    {
+    public:
+        ColorCorrectionModel::Impl* ccmLoss;
+        LossFunction(ColorCorrectionModel::Impl* ccm)
+            : ccmLoss(ccm) {};
+
+        /** @brief Reset dims to ccm->shape.
+        */
+        int getDims() const CV_OVERRIDE
+        {
+            return ccmLoss->shape;
+        }
+
+        /** @brief Reset calculation.
+        */
+        double calc(const double* x) const CV_OVERRIDE
+        {
+            Mat ccm_(ccmLoss->shape, 1, CV_64F);
+            for (int i = 0; i < ccmLoss->shape; i++)
+            {
+                ccm_.at<double>(i, 0) = x[i];
+            }
+            ccm_ = ccm_.reshape(0, ccmLoss->shape / 3);
+            return ccmLoss->calcLoss(ccm_);
+        }
+    };
+};
+
+ColorCorrectionModel::Impl::Impl()
+    : cs(*GetCS::getInstance().getRgb(COLOR_SPACE_SRGB))
+    , csEnum(COLOR_SPACE_SRGB)
+    , ccmType(CCM_LINEAR)
+    , distance(DISTANCE_CIE2000)
+    , linearizationType(LINEARIZATION_GAMMA)
+    , weights(Mat())
+    , gamma(2.2)
+    , deg(3)
+    , saturatedThreshold({ 0, 0.98 })
+    , initialMethodType(INITIAL_METHOD_LEAST_SQUARE)
+    , weightsCoeff(0)
+    , maxCount(5000)
+    , epsilon(1.e-4)
+{}
+
+Mat ColorCorrectionModel::Impl::prepare(const Mat& inp)
+{
+    switch (ccmType)
+    {
+    case cv::ccm::CCM_LINEAR:
+        shape = 9;
+        return inp;
+    case cv::ccm::CCM_AFFINE:
+    {
+        shape = 12;
+        Mat arr1 = Mat::ones(inp.size(), CV_64F);
+        Mat arr_out(inp.size(), CV_64FC4);
+        Mat arr_channels[3];
+        split(inp, arr_channels);
+        merge(std::vector<Mat> { arr_channels[0], arr_channels[1], arr_channels[2], arr1 }, arr_out);
+        return arr_out;
+    }
+    default:
+        CV_Error(Error::StsBadArg, "Wrong ccmType!");
+        break;
+    }
+}
+
+void ColorCorrectionModel::Impl::calWeightsMasks(const Mat& weightsList_, double weightsCoeff_, Mat saturateMask)
+{
+    // weights
+    if (!weightsList_.empty())
+    {
+        weights = weightsList_;
+    }
+    else if (weightsCoeff_ != 0)
+    {
+        pow(dst->toLuminant(cs.io), weightsCoeff_, weights);
+    }
+
+    // masks
+    Mat weight_mask = Mat::ones(src.rows, 1, CV_8U);
+    if (!weights.empty())
+    {
+        weight_mask = weights > 0;
+    }
+    this->mask = (weight_mask) & (saturateMask);
+
+    // weights' mask
+    if (!weights.empty())
+    {
+        Mat weights_masked = maskCopyTo(this->weights, this->mask);
+        weights = weights_masked / mean(weights_masked)[0];
+    }
+    maskedLen = (int)sum(mask)[0];
+}
+
+void ColorCorrectionModel::Impl::initialWhiteBalance(void)
+{
+    Mat schannels[4];
+    split(srcRgbl, schannels);
+    Mat dchannels[4];
+    split(dstRgbl, dchannels);
+    std::vector<double> initialVec = { sum(dchannels[0])[0] / sum(schannels[0])[0], 0, 0, 0,
+        sum(dchannels[1])[0] / sum(schannels[1])[0], 0, 0, 0,
+        sum(dchannels[2])[0] / sum(schannels[2])[0], 0, 0, 0 };
```

Comment:
No need to spit cv::Mat by channels for it. cv::sum supports channels: https://docs.opencv.org/5.x/d2/de8/group__core__array.html#ga716e10a2dd9e228e4d3c95818f106722